File: js/packages/go-bridge/ios/LoggerDriver.swift
//
//  Logger.swift
//  Berty
//
//  Created by Guilhem Fanton on 17/10/2018.
//

import os
import Bertybridge

enum LoggerError: Error {
  case emptyMessage
  case emptyNamespace
  case invalidLevel
}

enum Level: String {
  case debug = "DEBUG"
  case info = "INFO"
  case warn = "WARN"
  case error = "ERROR"
  case panic = "PANIC"
  case dPanic = "DPANIC"
  case fatal = "FATAL"
  case unknow = "UNKNOW"
}

enum Visibility {
  case visible
  case hidden
}

class LoggerDriver: NSObject, BertybridgeNativeLoggerDriverProtocol {
  var subsytem: String
  var category: String
  var scope: Visibility
  var isEnabled: Bool

  init(_ subsytem: String = "logger", _ category: String = "log") {
    self.subsytem = subsytem
    self.category = category
    self.scope = Visibility.visible
    self.isEnabled = true
  }

  func log(_ level: String?, namespace: String?, message: String?) throws {
    guard let ulevel = level, let level = Level(rawValue: ulevel) else {
      throw LoggerError.invalidLevel
    }

    guard let out = message else {
      throw LoggerError.emptyMessage
    }

    guard let subsystem = namespace else {
      throw LoggerError.emptyNamespace
    }

    if #available(iOS 10.0, *) {
        let logger = OSLog(subsystem: self.subsytem + "." + subsystem, category: self.category)

        var type: OSLogType
        switch level {
            // @FIXME(gfanton): on some device: debug log dont show up on the Console.
            // for the moment, use default type for debug
        case Level.debug:
            type = .default
        case Level.info:
            type = .info
        case Level.warn:
            type = .error
        case Level.error, Level.dPanic, Level.panic, Level.fatal:
            type = .fault
        default:
            type = .default
        }

        switch self.scope {
        case Visibility.visible: os_log("[%{public}@] %{public}@", log: logger, type: type, ulevel, out)
        case Visibility.hidden: os_log("[%{private}@] %{private}@", log: logger, type: type, ulevel, out)
        }
    } else {
        NSLog("[%@] [%@]: %@", level.rawValue, self.subsytem + "." + subsytem, out)
    }
  }

  func format(_ format: NSString, level: Level = Level.info, _ args: CVarArg...) {
      self.print(NSString(format: format, args), level: level)
  }

  func print(_ message: NSString, level: Level = Level.info, category: String? = nil) {
      let namespace = category ?? self.category
      do {
          try self.log(level.rawValue, namespace: namespace, message: message as String)
      } catch {
          NSLog("[%@] [%@]: %@", level.rawValue, self.subsytem + ".log", message)
      }
  }

  // @TODO: implement this
  open func levelEnabler(_ level: String?) -> Bool {
    return self.isEnabled
  }
}

File: js/android/app/src/main/java/tech/berty/gobridge/LoggerDriver.java
package tech.berty.gobridge;

import android.util.Log;
import android.util.LogPrinter;

import bertybridge.NativeLoggerDriver;

public class LoggerDriver implements NativeLoggerDriver {
    private final String subsystem;
    private final String category;
    private final boolean isEnabled;

    public LoggerDriver() {
        this("logger", "log");
    }

    public LoggerDriver(String subsystem) {
        this(subsystem, "log");
    }

    public LoggerDriver(String subsystem, String category) {
        this.subsystem = subsystem;
        this.category = category;
        this.isEnabled = true;
    }

    public void log(String level, String namespace, String message) throws Exception {
        LoggerLevel loggerLevel;

        try {
            loggerLevel = LoggerLevel.valueOf(level);
        } catch (Exception cause) {
            throw new Exception("Invalid level");
        }

        if (message == null) {
            throw new Exception("Empty message");
        }

        if (namespace != null && namespace != "") {
            namespace = this.subsystem + "." + namespace;
        } else {
            namespace = this.subsystem;
        }

        int priority;
        switch (loggerLevel) {
            case ASSERT:
                priority = Log.ASSERT;
                break;
            case DEBUG:
                priority = Log.DEBUG;
                break ;
            case PANIC:
            case DPANIC:
            case FATAL:
            case ERROR:
                priority = Log.ERROR;
                break ;
            case INFO:
                priority = Log.INFO;
                break ;
            case VERBOSE:
                priority = Log.VERBOSE;
                break ;
            case WARN:
                priority = Log.WARN;
                break ;
            case UNKNOW:
            default:
                priority = Log.INFO;
        }
        LogPrinter logPrinter = new LogPrinter(priority, namespace);
        String out = String.format("[%s] [%s]: %s", level, namespace, message);
        logPrinter.println(out);
    }

    public void format(String format, LoggerLevel loggerLevel, Object ... args) {
        this.print(String.format(format, args), loggerLevel, null);
    }

    public void print(String message, LoggerLevel loggerLevel, String category) {
        if (category == null) {
            category = this.category;
            try {
                this.log(loggerLevel.valueOf(), category, message);
            } catch (Exception e) {
                System.out.printf("[%s] [%s]: %s\n", loggerLevel.valueOf(), this.subsystem + ".log", message);
            }
        }
    }

    // @TODO: implement this
    public boolean levelEnabler(String level) {
        return this.isEnabled;
    }
}
